include/onnxruntime/core/graph/graph.h

```diff
@@ -795,6 +782,97 @@ class Graph {
   /** Returns the mutable parent graph if this is a subgraph */
   Graph* MutableParentGraph() { return parent_graph_; }
 
+  /** Sets the type of a NodeArg, replacing existing type/shape if any */
+  void SetNodeArgType(NodeArg& arg, const onnx::TypeProto& type_proto);
+
+  const Node* GetProducerNode(const std::string& node_arg_name) const {
+    auto iter = node_arg_to_producer_node_.find(node_arg_name);
+
+    if (iter != node_arg_to_producer_node_.end()) {
+      auto node_index = iter->second;
+      return GetNode(node_index);
+    }
+    return nullptr;
+  }
+
+  Node* GetMutableProducerNode(const std::string& node_arg_name) {
+    return const_cast<Node*>(GetProducerNode(node_arg_name));
```

Comment:
[https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es50-dont-cast-away-const](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es50-dont-cast-away-const) has some further discussion on this approach. If binary size is a primary concern, this may be fine. 

---
In reply to: [396855459](https://github.com/microsoft/onnxruntime/pull/3174#discussion_r396855459) [](ancestors = 396855459)